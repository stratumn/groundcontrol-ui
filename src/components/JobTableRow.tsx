// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import graphql from "babel-plugin-relay/macro";
import { Link } from "found";
import React, { Fragment } from "react";
import {
  Button,
  Table,
 } from "semantic-ui-react";

import { createFragmentContainer } from "react-relay";

import { JobTableRow_item } from "./__generated__/JobTableRow_item.graphql";

import Moment from "react-moment";
import RepositoryShortName from "./RepositoryShortName";

import "./JobTableRow.css";

const dateFormat = "L LTS";

export interface IProps {
  item: JobTableRow_item;
  onStop: (values: IProps) => any;
}

export function JobTableRow(props: IProps) {
  const {
    item: {
      name,
      owner,
      createdAt,
      updatedAt,
      priority,
      status,
    },
    onStop,
  } = props;
  const buttons: JSX.Element[] = [];

  let workspaceSlug = "-";
  let workspaceName = "-";
  let projectRepository = "-";
  let projectReference = "-";

  switch (owner.__typename) {
  case "Workspace":
    workspaceSlug = owner.slug;
    workspaceName = owner.name;
    break;
  case "Project":
    workspaceSlug = owner.workspace.slug;
    workspaceName = owner.workspace.name;
    projectRepository = owner.repository;
    projectReference = owner.reference;
    break;
  }

  let workspaceEl: JSX.Element;

  if (workspaceName === "-") {
    workspaceEl = <Fragment>{workspaceName}</Fragment>;
  } else {
    workspaceEl = (
      <Link to={`/workspaces/${workspaceSlug}`}>
        {workspaceName}
      </Link>
    );
  }

  if (status === "RUNNING") {
    const handleStop = () => onStop({ ...props });
    buttons.push((
      <Button
        key="stop"
        size="mini"
        compact={true}
        icon="stop"
        content="Stop"
        onClick={handleStop}
      />
    ));
  }

  return (
    <Table.Row className="JobTableRow">
      <Table.Cell>{name}</Table.Cell>
      <Table.Cell>
        {workspaceEl}
      </Table.Cell>
      <Table.Cell>
        <RepositoryShortName repository={projectRepository} />
      </Table.Cell>
      <Table.Cell>{projectReference}</Table.Cell>
      <Table.Cell>
        <Moment format={dateFormat}>{createdAt}</Moment>
      </Table.Cell>
      <Table.Cell>
        <Moment format={dateFormat}>{updatedAt}</Moment>
      </Table.Cell>
      <Table.Cell>{priority}</Table.Cell>
      <Table.Cell
        positive={status === "DONE"}
        warning={status === "RUNNING"}
        error={status === "FAILED"}
      >
        {status}
      </Table.Cell>
      <Table.Cell className="JobTableRowActions">
        {buttons}
      </Table.Cell>
    </Table.Row>
  );
}

export default createFragmentContainer(JobTableRow, graphql`
  fragment JobTableRow_item on Job {
    id
    name
    createdAt
    updatedAt
    owner {
      __typename
      ... on Workspace {
        slug
        name
      }
      ... on Project {
        repository
        reference
        workspace {
          slug
          name
        }
      }
    }
    priority
    status
  }`,
);
