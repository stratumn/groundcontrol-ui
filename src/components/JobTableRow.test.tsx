// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from "enzyme";
import React from "react";
import { mocked } from "ts-jest/utils";

import { JobTableRow_item } from "./__generated__/JobTableRow_item.graphql";

import { mockQueryPropAttrs } from "../testing/relay";

import { IProps, JobTableRow } from "./JobTableRow";

const other: JobTableRow_item = {
  ...mockQueryPropAttrs(),
  createdAt: "createdAt",
  id: "id",
  name: "name",
  owner: {
    __typename: "%other",
  },
  priority: "NORMAL",
  status: "QUEUED",
  updatedAt: "updatedAt",
};

const workspace: JobTableRow_item = {
  ...mockQueryPropAttrs(),
  createdAt: "createdAt",
  id: "id",
  name: "name",
  owner: {
    __typename: "Workspace",
    name: "workspaceName",
    slug: "workspaceSlug",
  },
  priority: "NORMAL",
  status: "QUEUED",
  updatedAt: "updatedAt",
};

const project: JobTableRow_item = {
  ...mockQueryPropAttrs(),
  createdAt: "createdAt",
  id: "id",
  name: "name",
  owner: {
    __typename: "Project",
    reference: "reference",
    repository: "repository",
    workspace: {
      name: "workspaceName",
      slug: "workspaceSlug",
    },
  },
  priority: "NORMAL",
  status: "QUEUED",
  updatedAt: "updatedAt",
};

const otherProps = {
  item: other,
  onStop: jest.fn(),
};

const workspaceProps = {
  item: workspace,
  onStop: jest.fn(),
};

const projectProps = {
  item: project,
  onStop: jest.fn(),
};

beforeEach(() => {
  mocked(otherProps.onStop).mockClear();
});

describe("<JobTableRow />", () => {

  it("renders correctly", () => {
    const wrapper = shallow(<JobTableRow {...otherProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when the owner is a workspace", () => {
    const wrapper = shallow(<JobTableRow {...workspaceProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when the owner is a project", () => {
    const wrapper = shallow(<JobTableRow {...projectProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("sets the positive attribute of the status cell to true when the job is done", () => {
    const wrapper = shallow(
      <JobTableRow
        {...{...otherProps, item: { ...other, status: "DONE" } }}
      />,
    );
    expect(wrapper.find("[positive=true]")).toHaveLength(1);
  });

  it("sets the warning attribute of the status cell to true when the job is running", () => {
    const wrapper = shallow(
      <JobTableRow
        {...{...otherProps, item: { ...other, status: "RUNNING" } }}
      />,
    );
    expect(wrapper.find("[warning=true]")).toHaveLength(1);
  });

  it("sets the error attribute of the status cell to true when the job failed", () => {
    const wrapper = shallow(
      <JobTableRow
        {...{...otherProps, item: { ...other, status: "FAILED" } }}
      />,
    );
    expect(wrapper.find("[error=true]")).toHaveLength(1);
  });

  it("shows a stop button when the job is running", () => {
    const wrapper = shallow(
      <JobTableRow
        {...{...otherProps, item: { ...other, status: "RUNNING" } }}
      />,
    );
    expect(wrapper.find("Button[icon='stop']")).toHaveLength(1);
  });

  it("doesn't show a stop button when the job isn't running", () => {
    const wrapper = shallow(
      <JobTableRow
        {...{...otherProps, item: { ...other, status: "DONE" } }}
      />,
    );
    expect(wrapper.find("Button[icon='stop']")).toHaveLength(0);
  });

  it("triggers onStop when the stop button is clicked", () => {
    const props: IProps = { ...otherProps, item: { ...other, status: "RUNNING" } };
    const wrapper = shallow(<JobTableRow {...props} />);
    wrapper.find("Button[icon='stop']").simulate("click");
    expect(otherProps.onStop).toBeCalledTimes(1);
    expect(otherProps.onStop).toBeCalledWith(props);
  });

});
