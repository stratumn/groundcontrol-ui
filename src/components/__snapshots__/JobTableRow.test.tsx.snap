// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JobTableRow /> renders correctly 1`] = `
<TableRow
  as="tr"
  cellAs="td"
  className="JobTableRow"
>
  <TableCell
    as="td"
  >
    name
  </TableCell>
  <TableCell
    as="td"
  >
    -
  </TableCell>
  <TableCell
    as="td"
  >
    <_default
      repository="-"
    />
  </TableCell>
  <TableCell
    as="td"
  >
    -
  </TableCell>
  <TableCell
    as="td"
  >
    <Moment
      format="L LTS"
    >
      createdAt
    </Moment>
  </TableCell>
  <TableCell
    as="td"
  >
    <Moment
      format="L LTS"
    >
      updatedAt
    </Moment>
  </TableCell>
  <TableCell
    as="td"
  >
    NORMAL
  </TableCell>
  <TableCell
    as="td"
    error={false}
    positive={false}
    warning={false}
  >
    QUEUED
  </TableCell>
  <TableCell
    as="td"
    className="JobTableRowActions"
  />
</TableRow>
`;

exports[`<JobTableRow /> renders correctly when the owner is a project 1`] = `
<TableRow
  as="tr"
  cellAs="td"
  className="JobTableRow"
>
  <TableCell
    as="td"
  >
    name
  </TableCell>
  <TableCell
    as="td"
  >
    <Link
      to="/workspaces/workspaceSlug"
    >
      workspaceName
    </Link>
  </TableCell>
  <TableCell
    as="td"
  >
    <_default
      repository="repository"
    />
  </TableCell>
  <TableCell
    as="td"
  >
    reference
  </TableCell>
  <TableCell
    as="td"
  >
    <Moment
      format="L LTS"
    >
      createdAt
    </Moment>
  </TableCell>
  <TableCell
    as="td"
  >
    <Moment
      format="L LTS"
    >
      updatedAt
    </Moment>
  </TableCell>
  <TableCell
    as="td"
  >
    NORMAL
  </TableCell>
  <TableCell
    as="td"
    error={false}
    positive={false}
    warning={false}
  >
    QUEUED
  </TableCell>
  <TableCell
    as="td"
    className="JobTableRowActions"
  />
</TableRow>
`;

exports[`<JobTableRow /> renders correctly when the owner is a workspace 1`] = `
<TableRow
  as="tr"
  cellAs="td"
  className="JobTableRow"
>
  <TableCell
    as="td"
  >
    name
  </TableCell>
  <TableCell
    as="td"
  >
    <Link
      to="/workspaces/workspaceSlug"
    >
      workspaceName
    </Link>
  </TableCell>
  <TableCell
    as="td"
  >
    <_default
      repository="-"
    />
  </TableCell>
  <TableCell
    as="td"
  >
    -
  </TableCell>
  <TableCell
    as="td"
  >
    <Moment
      format="L LTS"
    >
      createdAt
    </Moment>
  </TableCell>
  <TableCell
    as="td"
  >
    <Moment
      format="L LTS"
    >
      updatedAt
    </Moment>
  </TableCell>
  <TableCell
    as="td"
  >
    NORMAL
  </TableCell>
  <TableCell
    as="td"
    error={false}
    positive={false}
    warning={false}
  >
    QUEUED
  </TableCell>
  <TableCell
    as="td"
    className="JobTableRowActions"
  />
</TableRow>
`;
